CASSANDRA-19987 - Add Direct IO support for compaction reads

samueldlightfoot opened the pull request:
Status: merged

This work adds Direct IO support specifically for compaction reads of compressed tables. In its current state it should be easily extendable to be used by other bulk operations which would benefit from bypassing the page cache (and thus not polluting it).

It introduces the ability to perform SSTable scans using either the default open data file, or an ephemeral data file opened using DIO for the lifetime of the scan. This could be extended to use any alternative disk access mode in the future, if required.

patch by Sam Lightfoot; to be reviewed by Ariel Weisberg & Maxwell Guo for CASSANDRA-19987

https://issues.apache.org/jira/browse/CASSANDRA-19987

Maxwell-Guo commented on src/java/org/apache/cassandra/config/Config.java:
```diff
@@ -1245,6 +1246,12 @@ public enum DiskAccessMode
 
+    public enum ScanDiskAccessMode
+    {
+        disk_default,
```
why we do not use DiskAccessMode which has all the disk access mode?

samueldlightfoot replied:
This enum was intended to differentiate scans using the default disk access mode (conf.disk_access_mode via the existing dFile handle) from scans requiring a new, direct file handle. 

Update: code has now been updated to use DiskAccessMode. I think this integrates well with the existing setup.

Maxwell-Guo commented on src/java/org/apache/cassandra/config/DatabaseDescriptor.java:
```diff
@@ -901,6 +903,10 @@ else if (conf.repair_session_space.toMebibytes() > (int) (Runtime.getRuntime().m
         applyReadThresholdsValidations(conf);
 
+        initializeCompactionScanDiskAccessMode();
+        if (compactionScanDiskAccessMode != conf.compaction_scan_disk_access_mode)
```
why we don't just logging the compaction scan access mode ?  we can remove the if judegment.

Maxwell-Guo commented on src/java/org/apache/cassandra/io/util/ChannelProxy.java:
```diff
@@ -40,30 +42,55 @@
+        switch (ioMode)
+        {
+            case DIRECT:
+                return new OpenOption[]{ StandardOpenOption.READ, ExtendedOpenOption.DIRECT };
```
Is ExtendedOpenOption.DIRECT flag enough for direct io ?

samueldlightfoot replied:
Yes - confirmed with blktrace and async-prof shows bypassing of the page cache.

Maxwell-Guo replied:
What I mean is that does DIRECT flag is enough to ensure that all data (including metadata) sync to disk when crash happens. see 
`To provide this guarantee, the application must use fsync, or set the O_SYNC or O_DSYNC flag on the file descriptor via fcntl`
from 
https://archive.kernel.org/oldwiki/ext4.wiki.kernel.org/index.php/Clarifying_Direct_IO's_Semantics.html

samueldlightfoot replied:
Thanks for the links. This definitely seems applicable to the write path, but ChannelProxy is for a read only channel, which I do not believe requires the O_SYNC flag or similar, as there's no control data to flush.

We will currently still perform compaction writes with buffered IO (for now).  However, and I think you've already spotted this, the commit log DIO writes will likely need to be verified that meta is flushed on block allocation.

aweisberg replied:
Uh... the commit log is not safe as currently written. It writes to the file, but doesn't sync the metadata on flush. That means the commit log may claim it has flushed the data, but the filesystem journal has not been flushed so the file length will be wrong and could truncate the file on restart.

Additionally Direct IO doesn't actually make data durable on disk (emit write barriers) it just flushes it to the cache of the disk. If your disk cache is volatile then you can lose data.

samueldlightfoot replied:
Patched by Maxwell in CASSANDRA-20692 - resolving.

aweisberg commented on src/java/org/apache/cassandra/config/DatabaseDescriptor.java:
```diff
@@ -1695,6 +1699,31 @@ public static void applyPartitioner(Config conf)
+        }
+        else if (DiskAccessMode.direct == compactionScanDiskAccessMode)
+        {
+            if (conf.disk_optimization_strategy == Config.DiskOptimizationStrategy.ssd)
```
Should this be a limitation we enforce? We support read ahead. Maybe the user should be the one to decide if they want to have enough read ahead to support spinning disk with direct IO instead of having the kernel manage read ahead using OS settings?

aweisberg commented on src/java/org/apache/cassandra/config/DatabaseDescriptor.java:
```diff
@@ -3242,6 +3283,11 @@ public static void initializeCommitLogDiskAccessMode()
 
+    public static void initializeCompactionScanDiskAccessMode()
+    {
+        compactionScanDiskAccessMode = resolveCompactionScanDiskAccessMode(conf.disk_access_mode, conf.compaction_scan_disk_access_mode);
```
Inline `resolveCompactionScanDiskAccessMode` here?

aweisberg commented on src/java/org/apache/cassandra/db/compaction/AbstractCompactionStrategy.java:
```diff
@@ -581,4 +583,10 @@ public boolean supportsEarlyOpen()
         return true;
     }
+
+    protected static DiskAccessMode compactionScanDiskAccessMode()
```
Does this need to wrap `DatabaseDescriptor`? Just have everything static import and call `getCompactionScanDiskAccessMode`.

aweisberg commented on src/java/org/apache/cassandra/io/sstable/format/SSTableReader.java:
```diff
@@ -472,6 +475,9 @@ protected SSTableReader(Builder<?, ?> builder, Owner owner)
         this.sstableMetadata = builder.getStatsMetadata();
         this.header = builder.getSerializationHeader();
         this.dfile = builder.getDataFile();
+        this.directIOSupported = FileUtils.isDirectIOSupported(dfile.file())
```
Just curious in what situations does a `FileStore` not report a block size?

samueldlightfoot replied:
On digging a little deeper, it's various flavours of filesystems such as virtual/in-memory and network which can either throw or return zero values.

  1. Virtual/In-Memory Filesystems: According to https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/nio/file/FileStore.html, the default implementation can throw UnsupportedOperationException 
  for non-local storage devices. Some virtual filesystems (like https://github.com/google/jimfs/issues/183) may not properly implement getBlockSize().
  2. POSIX Specification: Per the https://man7.org/linux/man-pages/man3/statvfs.3.html, "It is unspecified whether all members of the returned struct have meaningful values on all file systems."
  3. Network Filesystems: NFS and other network filesystems may return unreliable or zero values for block size in certain conditions, as https://access.redhat.com/solutions/18357.

samueldlightfoot replied:
Comment added to FileUtils.isDirectIOSupported.

aweisberg commented on src/java/org/apache/cassandra/io/sstable/format/SSTableReader.java:
```diff
@@ -1064,12 +1070,17 @@ public ISSTableScanner getScanner(Range<Token> range)
         PartitionPositionBounds fullRange = getPositionsForFullRange();
         if (fullRange != null)
-            return new SSTableSimpleScanner(this, Collections.singletonList(fullRange));
+            return new SSTableSimpleScanner(this, Collections.singletonList(fullRange), diskAccessMode);
```
Can the scanner get the disk access mode from the `dfile` or `SSTableReader` directly instead of passing it in as a parameter since `this` is passed in?

That would simplify a little.

samueldlightfoot replied:
Not if the disk access mode we want to use for the scan is different to the default (non-scan) disk access mode of the data file (dFile).

I originally played with the idea of having another FileHandle on the SSTableReader specifically for scans, but since we currently only support **compaction** scans using a different disk access mode (for now) it didn't work particularly well (and creates a permanent fd rather than ephemeral for the lifetime of the scan).

aweisberg commented on src/java/org/apache/cassandra/io/sstable/format/SSTableReader.java:
```diff
@@ -1402,7 +1418,27 @@ public RandomAccessReader openDataReader()
+    {
+        if (diskAccessMode == dfile.diskAccessMode())
+        {
+            return dfile.createReaderForScan(OnReaderClose.RETAIN_FILE_OPEN);
```
Since the consequence of this and the next condition are the same maybe they should be the same condition? You could assign each condition to a local boolean and then have a single `if` on both booleans.

aweisberg commented on src/java/org/apache/cassandra/io/sstable/format/SSTableReader.java:
```diff
@@ -1402,7 +1418,27 @@ public RandomAccessReader openDataReader()
+        }
+
+        //noinspection resource - The FileHandle lifecycle is managed by the returned RandomAccessReader
+        FileHandle dataFile = dfile.toBuilder()
```
Exception safety leaking a file handle? It looks like we are just not exception safe in general around opening files looking at the callers of `FileHandle.complete`.

samueldlightfoot replied:
Added exception handling and resolved two other usages of complete without handling.

aweisberg commented on src/java/org/apache/cassandra/io/util/CompressedChunkReader.java:
```diff
@@ -119,6 +119,52 @@ default void close()
+
+        private final ChannelProxy channel;
+        private final int blockSize;
+        private final DirectThreadLocalByteBufferHolder bufferHolder;
```
We need to think through carefully the memory management here since we allocate these readers for each scan and then discard them immediately. We want the memory dedicated to this thread local to be returned promptly once the file handle is closed.

samueldlightfoot replied:
ByteBuffer cleaning added on close for all readers (direct & standard)

aweisberg commented on src/java/org/apache/cassandra/io/util/RandomAccessReader.java:
```diff
@@ -319,6 +319,34 @@ public void close()
+            }
+            finally
+            {
+                if (this.fileHandle != null)
```
When would we create a reader with a null file handle? Should that be a precondition in the constructor?

aweisberg commented on src/java/org/apache/cassandra/io/util/ThreadLocalReadAheadBuffer.java:
```diff
@@ -51,23 +53,26 @@ protected Map<String, Block> initialValue()
-     */
     public int remaining()
     {
+        assert hasBuffer();
```
This could be a `checkState`. I think we are at the point mostly where we don't use `assert` so often because no one turns them off anyways. If it's going to be an `assert` it should have a message.

aweisberg commented on src/java/org/apache/cassandra/io/util/ThreadLocalReadAheadBuffer.java:
```diff
@@ -51,23 +53,26 @@ protected Map<String, Block> initialValue()
     private final long channelSize;
 
     public ThreadLocalReadAheadBuffer(ChannelProxy channel, int bufferSize, BufferType bufferType)
```
Consider removing buffer size and then just fetch it from the buffer?

aweisberg commented on src/java/org/apache/cassandra/io/util/DirectThreadLocalReadAheadBuffer.java:
```diff
@@ -0,0 +1,53 @@
+    @Override
+    protected void loadBlock(ByteBuffer blockBuffer, long blockPosition, int sizeToRead)
+    {
+        int alignedSizeToRead = BitUtil.align(sizeToRead, blockSize);
```
When reading is it only required to aligned the end of the read to a block? Does the start of the read also need to be blocked aligned?

samueldlightfoot replied:
The start position is aligned in org.apache.cassandra.io.util.ThreadLocalReadAheadBuffer#fill 

L108: _long blockPosition = blockNo * (long) bufferSize;_

aweisberg commented on test/unit/org/apache/cassandra/io/util/StandardCompressedChunkReaderTest.java:
```diff
@@ -0,0 +1,116 @@
+        var paramsGen = compressionParams(Gens.constant(CompressionParams.DEFAULT_CHUNK_LENGTH));
+        var lengthGen = Gens.longs().between(1, 1 << 16);
+
+        qt().withSeed(-1871070464864118891L).forAll(Gens.random(), optionGen, paramsGen).check((rs, option, params) -> {
```
Hard coded seed intentional?

aweisberg commented on test/unit/org/apache/cassandra/io/util/CompressedChunkReaderTest.java:
```diff
@@ -20,122 +20,50 @@
-                case Lz4: return CompressionParams.lz4(chunkLengths.next(rs));
-                case Zstd: return CompressionParams.zstd(chunkLengths.next(rs));
-                default: throw new UnsupportedOperationException(kind.name());
+                case Noop:
```
Try and avoid style churn

aweisberg commented on test/unit/org/apache/cassandra/io/util/CompressedChunkReaderTest.java:
```diff
@@ -20,122 +20,50 @@
-                                           .finishOnClose(false)
-                                           .bufferSize(bufferSizes.next(rs))
-                                           .build();
+        Noop, Snappy, Deflate, Lz4, Zstd
```
Style churn

aweisberg commented on src/java/org/apache/cassandra/io/util/FileHandle.java:
```diff
@@ -322,13 +392,14 @@ public Builder withCrcCheckChance(Supplier<Double> crcCheckChanceSupplier)
         {
-            this.mmapped = mmapped;
+            if (mmapped)
+                withDiskAccessMode(DiskAccessMode.mmap);
```
Should this have an `else` for the false condition? Otherwise remove the parameter and have it always switch to mmaped.

aweisberg commented on src/java/org/apache/cassandra/io/util/FileHandle.java:
```diff
@@ -62,17 +70,41 @@ public class FileHandle extends SharedCloseableImpl
+    private final DiskAccessMode diskAccessMode;
+
+    // Properties to support unbuilding via toBuilder
+    private final ChunkCache chunkCache;
```
I don't love adding all this to support unbuild. Maybe just keep the builder and allow `complete` to be called multiple times? That is not thread safe thought.

I don't see a much better way to do it at the moment.

samueldlightfoot replied:
I agree, however I'm not sure on a better alternative. Originally I had a new dataFileFactory function variable on the SSTableReader that allowed to modify the diskAccessMode, but it sprawled more than adding the unbuild support as it needed all usaged to set the new prop.

Curious if you have any other thoughts.

aweisberg reviewed (changes requested):
I don't have much substantial feedback. I think I need to do another pass on `CompressedChunkReader` to look for the unknowns, but everything looks reasonable.

I think a good test would be test trying to read from > 4gb files to verify that everything works end to end when the file position is large.

samueldlightfoot committed:
Add Direct IO support for compaction reads

This change adds support for using O_DIRECT (direct I/O) when reading SSTables
during compaction. Direct I/O bypasses the OS page cache, which can be beneficial
for compaction workloads where data is typically read once and not accessed again
soon after.

Key changes:
- Add DiskAccessMode.direct option for scan operations
- Introduce DirectThreadLocalByteBufferHolder and DirectThreadLocalReadAheadBuffer
  for aligned buffer management required by O_DIRECT
- Add ByteBufferHolder interface to abstract buffer management
- Modify FileHandle to support building direct I/O readers via toBuilder()
- Add startup check to verify Direct IO support on configured data directories
- Enable read-ahead for uncompressed data in Direct CompressedChunkReader
- Move direct IO support logic into FileHandle (from SSTableReader)
- Add comprehensive tests for direct I/O chunk readers and buffer holders

patch by Sam Lightfoot; reviewed by Ariel Weisberg for CASSANDRA-19987

aweisberg commented on src/java/org/apache/cassandra/io/util/DirectThreadLocalByteBufferHolder.java:
```diff
@@ -0,0 +1,79 @@
+ */
+public final class DirectThreadLocalByteBufferHolder implements ByteBufferHolder
+{
+    private static final ConcurrentHashMap<Integer, FastThreadLocal<ByteBuffer>> buffersByBlockSize = new ConcurrentHashMap<>();
```
Does it make sense to use an integer when any sufficiently large buffer in a given thread is sufficient? Shouldn't it be more like the thread local one where it just keeps allocating bigger buffers per thread?

samueldlightfoot replied:
That would work if we tracked the max block size seen instead of the Map, although I think the Map is a little more verbose about the alignment requirements of each known block size.

aweisberg reviewed (approved):
+1 TY, we can add support for cursor based compaction in [CASSANDRA-21147](https://issues.apache.org/jira/browse/CASSANDRA-21147)

Maxwell-Guo commented on src/java/org/apache/cassandra/config/DatabaseDescriptor.java:
```diff
@@ -1763,7 +1780,7 @@ private static Pair<DiskAccessMode, Boolean> resolveCommitLogWriteDiskAccessMode
                 File commitLogLocationDir = new File(commitLogLocation);
                 PathUtils.createDirectoriesIfNotExists(commitLogLocationDir.toPath());
-                directIOSupported = FileUtils.getBlockSize(commitLogLocationDir) > 0;
+                directIOSupported = FileUtils.isDirectIOSupported(commitLogLocationDir);
```
why you change the code here ?

samueldlightfoot replied:
To ensure the direct IO compatibility checks are consistent throughout

Maxwell-Guo commented on src/java/org/apache/cassandra/io/util/FileUtils.java:
```diff
@@ -740,14 +741,54 @@ public static void deleteDirectoryIfEmpty(Path path) throws IOException
+            if (blockSize(testFile) <= 0)
+                return false;
+
+            try (FileChannel channel = FileChannel.open(testFile.toPath(),
```
1、so you open the channel ? just to see if we can use direct io ? and but if the channel is successfully opend, when should we close it ?
2、 why the original method "blockSize(file) > 0" can not meet your need ?

samueldlightfoot replied:
The try-with resources pattern used here will close the channel immediately after opening.

https://manpages.ubuntu.com/manpages/xenial//man2/open.2.html

>        O_DIRECT support was added under Linux in kernel version 2.4.10.  Older Linux kernels simply ignore  this
>        flag.  Some filesystems may not implement the flag and open() will fail with EINVAL if it is used.

The blockSize check is not wrong, but it alone does not guarantee Direct I/O is supported by the underlying filesystem.

Maxwell-Guo replied:
you are right.

Maxwell-Guo reviewed (approved):
LGTM

samueldlightfoot commented:
[Committed](https://github.com/apache/cassandra/commit/6f5fe8c06d5831daae1f1a4f2412f3185798dca0)
